cmd/outline-ss-server/metrics.go

```diff
@@ -0,0 +1,228 @@
+			Help:      "Histogram of number of bytes from client to proxy, for detecting possible probes",
+		}, []string{"port", "status", "error"}),
+		tcpOpenConnections: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "shadowsocks",
+			Subsystem: "tcp",
+			Name:      "connections_opened",
+			Help:      "Count of open TCP connections",
+		}, []string{"location"}),
+		tcpClosedConnections: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "shadowsocks",
+			Subsystem: "tcp",
+			Name:      "connections_closed",
+			Help:      "Count of closed TCP connections",
+		}, []string{"location", "status", "access_key"}),
+		tcpConnectionDurationMs: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "shadowsocks",
+				Subsystem: "tcp",
+				Name:      "connection_duration_ms",
+				Help:      "TCP connection duration distributions.",
+				Buckets: []float64{
+					100,
+					float64(time.Second.Milliseconds()),
+					float64(time.Minute.Milliseconds()),
+					float64(time.Hour.Milliseconds()),
+					float64(24 * time.Hour.Milliseconds()),     // Day
+					float64(7 * 24 * time.Hour.Milliseconds()), // Week
+				},
+			}, []string{"status"}),
+		dataBytes: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "shadowsocks",
+				Name:      "data_bytes",
+				Help:      "Bytes transferred by the proxy, per access key",
+			}, []string{"dir", "proto", "access_key"}),
+		dataBytesPerLocation: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "shadowsocks",
+				Name:      "data_bytes_per_location",
+				Help:      "Bytes transferred by the proxy, per location",
+			}, []string{"dir", "proto", "location"}),
+		timeToCipherMs: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "shadowsocks",
+				Name:      "time_to_cipher_ms",
+				Help:      "Time needed to find the cipher",
+				Buckets:   []float64{0.1, 1, 10, 100, 1000},
+			}, []string{"proto", "found_key"}),
+		udpPacketsFromClientPerLocation: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "shadowsocks",
+				Subsystem: "udp",
+				Name:      "packets_from_client_per_location",
+				Help:      "Packets received from the client, per location and status",
+			}, []string{"location", "status"}),
+		udpAddedNatEntries: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: "shadowsocks",
+				Subsystem: "udp",
+				Name:      "nat_entries_added",
+				Help:      "Entries added to the UDP NAT table",
+			}),
+		udpRemovedNatEntries: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: "shadowsocks",
+				Subsystem: "udp",
+				Name:      "nat_entries_removed",
+				Help:      "Entries removed from the UDP NAT table",
+			}),
+	}
+
+	// TODO: Is it possible to pass where to register the collectors?
+	registerer.MustRegister(m.buildInfo, m.accessKeys, m.ports, m.tcpProbes, m.tcpOpenConnections, m.tcpClosedConnections, m.tcpConnectionDurationMs,
```

Comment:
Do we want to return an error instead of using `MustRegister` which might `panic` here? Cuz I think metrics feature is optional to run the server.

Reply:
We could, but the panic seems reasonable. Metrics are always colelcted, but may not be exposed if you don't specify the metrics endpoint. But it won't panic. It really only panics if the code is registering something invalid.

I think i'll keep it as is for now, since this is intended to be just a refactor.